### PR TITLE
Issue #SB-7863, Fix : MTF - When five options are given, one of them is always correctly matched

### PR DESCRIPTION
**Pull request by ramj-funtoot**
https://project-sunbird.atlassian.net/browse/SB-7863

Issue description
The RHS were shuffled randomly, but one or more options may end up being in the original index causing those options to get answered correctly without the student rearranging the options. This caused a non-zero (partial) score being awarded to the student.

Fix description
The fix is to shuffle the options such that no option is at its original index.

The Sattolo's Algorithm (based on Fisher-Yates shuffle algorithm) has been used in arriving at a shuffled array that does not have any element in its original index.

### CODE REVIEW

File: renderer/plugin.js [hunk 1]
```diff
@@ -115,7 +115,34 @@ org.ekstep.questionunitmtf.RendererPlugin = org.ekstep.contentrenderer.questionU
...
   },
   logTelemetryItemResponse: function (data) {
-    QSTelemetryLogger.logEvent(QSTelemetryLogger.EVENT_TYPES.RESPONSE, {"type": "INPUT", "values": data});
+    QSTelemetryLogger.logEvent(QSTelemetryLogger.EVENT_TYPES.RESPONSE, { "type": "INPUT", "values": data });
+  },
+  /**
+   * shuffles the options array
+   */
+  shuffleOptions: function (options) {
+    var shuffled = [];
```
Comment (vinukumar-vs) [inline renderer/plugin.js]: By using simple logic with _.reject & _.shuffle we can achieve this.

var arr = [1,2,3,4,5]
var shArr = _.shuffle(arr)
_.reject(arr, function(val, i){return val != shArr[i]})

if _.reject array length is 0, then all options are mismatched. Else do _shuffle again & try

Comment (vinukumar-vs) [reply renderer/plugin.js]: var arr = [1,2,3,4,5] => undefined
_.shuffle(arr) =>(5) [2, 3, 5, 4, 1]
_.shuffle(arr) => (5) [2, 5, 4, 1, 3]

var shArr = _.shuffle(arr) => undefined
shArr => (5) [1, 5, 4, 3, 2]
_.reject(arr, function(val, i){return val != shArr[i]}) => [1]
var shArr = _.shuffle(arr) => undefined
_.reject(arr, function(val, i){return val != shArr[i]}) => []

File: renderer/plugin.js [hunk 2]
```diff
@@ -115,7 +115,34 @@ org.ekstep.questionunitmtf.RendererPlugin = org.ekstep.contentrenderer.questionU
...
+    });
+    return shuffled;
+  },
+  /**
+   * deranges (shuffles such that no element will remain in its original index) 
+   * the elements the given array. This is a JavaScript implementation of
+   * Sattolo's algorithm [https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle#Sattolo's_algorithm]
+   */
+  derange: function (array) {
+    var m = array.length, t, i;
```
[the remark this answers is not in the document]
Comment (vinukumar-vs) [reply renderer/plugin.js]: You can remove this function completely by using above comment to handle option mismatch.

File: renderer/plugin.js (hunk 1 above)
Comment (ramj-funtoot) [reply renderer/plugin.js]: @vinukumar-vs I actually compared the logic I wrote against the `_.shuffle()` and I found that the chances for `_.shuffle()` to return array with elements in the same index is very high. So, there is no guarantee that we will get the desired outcome in 1 or 2 iterations. 

- 

-

**Pull request merged**